cum/scrapers/mangadex.py

```diff
@@ -0,0 +1,143 @@
+            url_match = re.search(MangadexChapter.url_re, url)
+            if not url_match:
+                continue
+            name = a.string
+            name = name.strip()
+            name_parts = re.search(self.name_re, name)
+            chapter = name_parts.group(1) if name_parts else name
+            title = name_parts.group(2) if name_parts else None
+            title = None if title == "Read Online" else title
+            # TODO add an option to filter by language
+            language = a.parent.parent.find('img', src=self.language_re)['title']
+            groups = [a.parent.parent.find('a', href=self.group_re).string]
+            c = MangadexChapter(name=manga_name, alias=self.alias,
+                                chapter=chapter,
+                                url='https://mangadex.com' + url,
+                                groups=groups, title=title)
+            chapters = [c] + chapters
+        return chapters
+
+    @property
+    def name(self):
+        title_re = re.compile(r'^(.+) \(Manga\) - MangaDex')
+        title = self.soup.find('title').string.strip()
+        return re.search(title_re, title).group(1)
+
+
+class MangadexChapter(BaseChapter):
+    url_re = re.compile(r'(?:https?://mangadex\.com)?/chapter/([0-9]+)')
+    # There is an inlined js with a bunch of useful variables
+    hash_re = re.compile(r'var dataurl ?= ?\'([A-Za-z0-9]{32})\'')
+    # Example: (mind that the trailing comma is invalid json)
+    # var page_array = [
+    # 'x1.jpg','x2.jpg','x3.jpg','x4.jpg','x5.jpg','x6.png',];
+    pages_re = re.compile(r'var page_array ?= ?\[([^\]]+)\]', re.DOTALL)
+    # Just extract the single page name like: x1.jpg
+    single_page_re = re.compile(r'\s?\'([^\']+)\',?')
+    # This can be a mirror server or data path. Example:
+    # var server = 'https://s2.mangadex.com/'
+    # var server = '/data/'
+    server_re = re.compile(r'var server ?= ?\'([^\']+)\'')
+    uses_pages = True
+
+    @staticmethod
+    def _reader_get(url, page_index):
+        return requests.get(url)
+
+    def available(self):
+        self.r = self.reader_get(1)
+        if not len(self.r.text):
+            return False
+        elif self.r.status_code == 404:
+            return False
+        elif re.search(re.compile(r'Chapter #[0-9]+ does not exist.'),
+                       self.r.text):
+            return False
+        else:
+            return True
+
+    def download(self):
+        if getattr(self, 'r', None):
+            r = self.r
+        else:
+            r = self.reader_get(1)
+        soup = BeautifulSoup(r.text, config.get().html_parser)
+        chapter_hash = re.search(self.hash_re, r.text).group(1)
+        pages_var = re.search(self.pages_re, r.text)
+        pages = re.findall(self.single_page_re, pages_var.group(1))
+        files = [None] * len(pages)
+        mirror = re.search(self.server_re, r.text).group(1)
+        server = 'https://mangadex.com' + mirror if mirror[0] == '/' else mirror
```

Comment:
Small nitpick: `urllib.parse.urljoin` may be more robust/elegant/pythonic/whatever